sklearn/neighbors/_nearest_centroid.py

```diff
@@ -255,7 +314,100 @@ def predict(self, X):
         """
         check_is_fitted(self)
 
-        X = self._validate_data(X, accept_sparse="csr", reset=False)
-        return self.classes_[
-            pairwise_distances_argmin(X, self.centroids_, metric=self.metric)
-        ]
+        if len(np.unique(self.class_priors_, return_counts=True)[0]) == 1:
+            X = self._validate_data(X, accept_sparse="csr", reset=False)
+            return self.classes_[
+                pairwise_distances_argmin(X, self.centroids_, metric=self.metric)
+            ]
+        else:
+            return super().predict(X)
+
+    def _decision_function(self, X):
+        check_is_fitted(self, "centroids_")
+
+        X = self._validate_data(X, reset=False, accept_sparse="csr")
+
+        # check if this works with sparse matrix formats later
+        discriminant_score = np.empty(
+            (X.shape[0], self.classes_.size), dtype=np.float64
+        )
+
+        for cur_class in range(self.classes_.size):
+            Xdist = X - self.centroids_[cur_class, :]
+            Xdist_norm = np.square(Xdist / self.within_class_std_)
+            # Hastie et al. (2009), p. 652, Eq. (18.2)
+            discriminant_score[:, cur_class] = np.squeeze(
+                -np.sum(Xdist_norm, axis=1)
+                + 2.0 * np.log(self.class_priors_[cur_class])
+            )
+
+        return discriminant_score
+
+    def decision_function(self, X):
```

Comment:
We can replace all the code below with:

```python
    def _check_euclidean_metric(self):
        return self.metric == "euclidean"

    decision_function = available_if(_check_euclidean_metric)(
        DiscriminantAnalysisPredictionMixin.decision_function
    )
    predict_proba = available_if(_check_euclidean_metric)(
        DiscriminantAnalysisPredictionMixin.predict_proba
    )
    predict_log_proba = available_if(_check_euclidean_metric)(
        DiscriminantAnalysisPredictionMixin.predict_log_proba
    )
```

Reply:
I'm not sure if `QuadraticDiscriminantAnalysis` supports sparse matrices if that's not the case, should we revert to the pattern of calling `super().decision_function(X)` in order to override the docstring as follows:

```
class DiscriminantAnalysisPredictionMixin:
    """Mixin class for QuadraticDiscriminantAnalysis and NearestCentroid."""

    def decision_function(self, X):
        """Apply decision function to an array of samples.

        Parameters
        ----------
        X : {array-like, sparse matrix} of shape (n_samples, n_features)
            Array of samples (test vectors).
```

Reply:
`QuadraticDiscriminantAnalysis` does not support sparse matrices. So they need different documentation.